File: app/lib/models/plan.dart
/// Dart mirror of the planner's output schema.
///
/// Keep this in sync with `server/src/schemas/plan.ts` — that file is the
/// source of truth, since it is what constrains the model's output.
library;

List<String> _strings(dynamic value) =>
    (value as List?)?.map((e) => e.toString()).toList() ?? const [];

enum PlanStatus { complete, needsInfo, infeasible }

PlanStatus planStatusFrom(String? value) => switch (value) {
      'needs_info' => PlanStatus.needsInfo,
      'infeasible' => PlanStatus.infeasible,
      _ => PlanStatus.complete,
    };

enum TimeOfDay { morning, afternoon, evening }

TimeOfDay timeOfDayFrom(String? value) => switch (value) {
      'afternoon' => TimeOfDay.afternoon,
      'evening' => TimeOfDay.evening,
      _ => TimeOfDay.morning,
    };

/// One budget category: what was intended against what the plan costs.
/// The Budget tab draws these as two stacked bars.
class CategoryBudget {
  const CategoryBudget({
    required this.name,
    required this.planned,
    required this.estimated,
  });

  final String name;
  final num planned;
  final num estimated;

  num get delta => estimated - planned;
  bool get isOver => delta > 0;

  /// "on plan" when the gap is small enough not to matter to the reader.
  bool get isOnPlan => delta.abs() < 1;

  factory CategoryBudget.fromJson(String name, Map<String, dynamic> json) =>
      CategoryBudget(
        name: name,
        planned: json['planned'] as num? ?? 0,
        estimated: json['estimated'] as num? ?? 0,
      );
}

class BudgetBreakdown {
  const BudgetBreakdown(this.categories);

  /// Ordered lodging → transport → food → activities → buffer, as designed.
  final List<CategoryBudget> categories;

  /// The largest single value across all categories, which every bar is drawn
  /// as a fraction of.
  num get peak => categories.fold<num>(
        0,
        (max, c) => [max, c.planned, c.estimated].reduce((a, b) => a > b ? a : b),
      );

  static const _order = ['lodging', 'transport', 'food', 'activities', 'buffer'];

  factory BudgetBreakdown.fromJson(Map<String, dynamic> json) => BudgetBreakdown([
        for (final name in _order)
          if (json[name] is Map<String, dynamic>)
            CategoryBudget.fromJson(name, json[name] as Map<String, dynamic>),
      ]);
}

class TripSummary {
  const TripSummary({
    required this.title,
    required this.destinations,
    required this.durationDays,
    required this.currency,
    required this.estimatedTotalCost,
    required this.overBudget,
    required this.budgetBreakdown,
    this.startDate,
    this.endDate,
    this.totalBudget,
    this.assumptions = const [],
  });

  final String title;
  final List<String> destinations;
  final String? startDate;
  final String? endDate;
  final num durationDays;
  final String currency;
  final num? totalBudget;
  final BudgetBreakdown budgetBreakdown;
  final num estimatedTotalCost;

  /// When true, the client must visually flag the budget section per the spec.
  final bool overBudget;
  final List<String> assumptions;

  factory TripSummary.fromJson(Map<String, dynamic> json) => TripSummary(
        title: json['title'] as String? ?? '',
        destinations: _strings(json['destinations']),
        startDate: json['start_date'] as String?,
        endDate: json['end_date'] as String?,
        durationDays: json['duration_days'] as num? ?? 0,
        currency: json['currency'] as String? ?? 'USD',
        totalBudget: json['total_budget'] as num?,
        budgetBreakdown: BudgetBreakdown.fromJson(
          json['budget_breakdown'] as Map<String, dynamic>? ?? const {},
        ),
        estimatedTotalCost: json['estimated_total_cost'] as num? ?? 0,
        overBudget: json['over_budget'] as bool? ?? false,
        assumptions: _strings(json['assumptions']),
      );
}

/// A conflict the planner surfaced rather than silently resolving. The spec is
/// explicit that these should be shown prominently, not buried.
class Conflict {
  const Conflict({
    required this.tag,
    required this.description,
    required this.membersInvolved,
    required this.resolution,
  });

  /// Pace / Mobility / Food / Budget / Interests — rendered as the card's pill.
  final String tag;
  final String description;
  final List<String> membersInvolved;
  final String resolution;

  factory Conflict.fromJson(Map<String, dynamic> json) => Conflict(
        tag: json['tag'] as String? ?? 'Pace',
        description: json['description'] as String? ?? '',
        membersInvolved: _strings(json['members_involved']),
        resolution: json['resolution'] as String? ?? '',
      );
}

class PlanBlock {
  const PlanBlock({
    required this.timeOfDay,
    required this.activity,
    required this.description,
    required this.location,
    required this.estimatedDurationMinutes,
    required this.suitedForMembers,
    required this.optional,
    this.estimatedCostPerPerson,
    this.weatherBackup,
  });

  final TimeOfDay timeOfDay;
  final String activity;
  final String description;
  final String location;
  final num estimatedDurationMinutes;
  final num? estimatedCostPerPerson;

  /// Member ids. Fewer than the full group means this is a split-group block.
  final List<String> suitedForMembers;
  final bool optional;
  final String? weatherBackup;

  factory PlanBlock.fromJson(Map<String, dynamic> json) => PlanBlock(
        timeOfDay: timeOfDayFrom(json['time_of_day'] as String?),
        activity: json['activity'] as String? ?? '',
        description: json['description'] as String? ?? '',
        location: json['location'] as String? ?? '',
        estimatedDurationMinutes: json['estimated_duration_minutes'] as num? ?? 0,
        estimatedCostPerPerson: json['estimated_cost_per_person'] as num?,
        suitedForMembers: _strings(json['suited_for_members']),
        optional: json['optional'] as bool? ?? false,
        weatherBackup: json['weather_backup'] as String?,
      );
}

class PlanDay {
  const PlanDay({
    required this.day,
    required this.location,
    required this.blocks,
    this.date,
    this.lodgingAreaSuggestion,
    this.notes,
  });

  final num day;
  final String? date;
  final String location;
  final String? lodgingAreaSuggestion;
  final List<PlanBlock> blocks;
  final String? notes;

  /// Per-person cost for the day — the sum of its blocks, as the design's
  /// "€75 pp planned" label. Optional blocks count only when shown.
  num costPerPerson({bool includeOptional = true}) => blocks
      .where((b) => includeOptional || !b.optional)
      .fold<num>(0, (sum, b) => sum + (b.estimatedCostPerPerson ?? 0));

  factory PlanDay.fromJson(Map<String, dynamic> json) => PlanDay(
        day: json['day'] as num? ?? 0,
        date: json['date'] as String?,
        location: json['location'] as String? ?? '',
        lodgingAreaSuggestion: json['lodging_area_suggestion'] as String?,
        blocks: (json['blocks'] as List? ?? const [])
            .map((b) => PlanBlock.fromJson(b as Map<String, dynamic>))
            .toList(),
        notes: json['notes'] as String?,
      );
}

class PlanMember {
  const PlanMember({required this.id, required this.name});

  final String id;
  final String name;

  factory PlanMember.fromJson(Map<String, dynamic> json) => PlanMember(
        id: json['id'] as String? ?? '',
        name: json['name'] as String? ?? '',
      );
}

/// A question the planner needs answered, rendered as a numbered card on the
/// Trip tab when the plan is [PlanStatus.needsInfo].
class ClarifyingQuestion {
  const ClarifyingQuestion({
    required this.id,
    required this.question,
    required this.why,
    required this.answerType,
    this.options = const [],
    this.placeholder,
  });

  final String id;
  final String question;

  /// What changes depending on the answer — shown under the question.
  final String why;

  /// 'choice' renders single-select chips; 'text' renders an input.
  final String answerType;
  final List<String> options;
  final String? placeholder;

  bool get isChoice => answerType == 'choice';

  factory ClarifyingQuestion.fromJson(Map<String, dynamic> json) =>
      ClarifyingQuestion(
        id: json['id'] as String? ?? '',
        question: json['question'] as String? ?? '',
        why: json['why'] as String? ?? '',
        answerType: json['answer_type'] as String? ?? 'text',
        options: _strings(json['options']),
        placeholder: json['placeholder'] as String?,
      );
}

/// The last planning attempt that produced nothing.
///
/// Deliberately worded against revisions rather than days: a planning run
/// either writes a revision or it doesn't, so "it planned days 1 and 2" is not
/// a state this system can be in. [lastGoodRevision] is what actually survived.
class PlanFailure {
  const PlanFailure({
    required this.reasonCode,
    required this.reason,
    required this.elapsedMs,
    required this.at,
    this.lastGoodRevision,
    this.lastGoodDays,
    this.userRequest,
  });

  final String reasonCode;

  /// One plain sentence, already worded by the server. Shown as-is.
  final String reason;
  final int elapsedMs;
  final DateTime at;

  /// The revision still on the trip, or null when nothing was ever planned.
  final int? lastGoodRevision;
  final int? lastGoodDays;

  /// The refinement being attempted, when the failure came from one.
  final String? userRequest;

  /// Whether there is a plan to fall back to. Drives which actions are offered.
  bool get hasFallback => lastGoodRevision != null;

  /// `STOPPED AT 14:32` — local time, since that's when the user saw it stop.
  String get stoppedAtLabel {
    final local = at.toLocal();
    return '${local.hour.toString().padLeft(2, '0')}:'
        '${local.minute.toString().padLeft(2, '0')}';
  }

  factory PlanFailure.fromJson(Map<String, dynamic> json) => PlanFailure(
        reasonCode: json['reason_code'] as String? ?? 'unknown',
        reason: json['reason'] as String? ?? 'The planner stopped',
        elapsedMs: (json['elapsed_ms'] as num?)?.toInt() ?? 0,
        at: DateTime.tryParse(json['at'] as String? ?? '') ?? DateTime.now(),
        lastGoodRevision: (json['last_good_revision'] as num?)?.toInt(),
        lastGoodDays: (json['last_good_days'] as num?)?.toInt(),
        userRequest: json['user_request'] as String?,
      );
}

class Plan {
  const Plan({
    required this.conversationalSummary,
    required this.status,
    required this.trip,
    this.missingInfo = const [],
    this.members = const [],
    this.conflicts = const [],
    this.itinerary = const [],
    this.packingAndPrepNotes = const [],
    this.verifyBeforeBooking = const [],
    this.clarifyingQuestions = const [],
  });

  /// Part A — the prose the user reads.
  final String conversationalSummary;
  final PlanStatus status;
  final List<String> missingInfo;
  final TripSummary trip;
  final List<PlanMember> members;
  final List<Conflict> conflicts;
  final List<PlanDay> itinerary;
  final List<String> packingAndPrepNotes;
  final List<String> verifyBeforeBooking;
  final List<ClarifyingQuestion> clarifyingQuestions;

  /// The spec's rule: on needs_info, show questions instead of an itinerary.
  bool get shouldRenderItinerary =>
      status != PlanStatus.needsInfo && itinerary.isNotEmpty;

  String nameForMember(String id) =>
      members.where((m) => m.id == id).map((m) => m.name).firstOrNull ?? id;

  factory Plan.fromJson(Map<String, dynamic> json) => Plan(
        conversationalSummary: json['conversational_summary'] as String? ?? '',
        status: planStatusFrom(json['status'] as String?),
        missingInfo: _strings(json['missing_info']),
        trip: TripSummary.fromJson(
          json['trip'] as Map<String, dynamic>? ?? const {},
        ),
        members: (json['members'] as List? ?? const [])
            .map((m) => PlanMember.fromJson(m as Map<String, dynamic>))
            .toList(),
        conflicts: (json['conflicts'] as List? ?? const [])
            .map((c) => Conflict.fromJson(c as Map<String, dynamic>))
            .toList(),
        itinerary: (json['itinerary'] as List? ?? const [])
            .map((d) => PlanDay.fromJson(d as Map<String, dynamic>))
            .toList(),
        packingAndPrepNotes: _strings(json['packing_and_prep_notes']),
        verifyBeforeBooking: _strings(json['verify_before_booking']),
        clarifyingQuestions: (json['clarifying_questions'] as List? ?? const [])
            .map((q) => ClarifyingQuestion.fromJson(q as Map<String, dynamic>))
            .toList(),
      );
}

/// A stored plan revision. Refinements produce a new revision rather than
/// mutating the last one, so the group can compare or roll back.
class PlanRevision {
  const PlanRevision({
    required this.id,
    required this.revision,
    required this.plan,
    required this.createdAt,
    this.userRequest,
  });

  final String id;
  final int revision;
  final Plan plan;
  final DateTime createdAt;
  final String? userRequest;

  factory PlanRevision.fromJson(Map<String, dynamic> json) => PlanRevision(
        id: json['id'] as String,
        revision: (json['revision'] as num?)?.toInt() ?? 1,
        plan: Plan.fromJson(json['plan'] as Map<String, dynamic>),
        createdAt:
            DateTime.tryParse(json['created_at'] as String? ?? '') ?? DateTime.now(),
        userRequest: json['user_request'] as String?,
      );
}

extension _FirstOrNull<T> on Iterable<T> {
  T? get firstOrNull => isEmpty ? null : first;
}

/// One bubble in the Refine tab. Derived server-side from plan revisions.
class ChatMessage {
  const ChatMessage({
    required this.role,
    required this.text,
    this.revision = 0,
  });

  final String role; // 'user' | 'bot'
  final String text;
  final int revision;

  bool get isUser => role == 'user';

  factory ChatMessage.fromJson(Map<String, dynamic> json) => ChatMessage(
        role: json['role'] as String? ?? 'bot',
        text: json['text'] as String? ?? '',
        revision: (json['revision'] as num?)?.toInt() ?? 0,
      );
}

/// The outcome of one Refine turn. Exactly one of [plan] or [reply] is set:
/// a link import answers conversationally, an instruction returns a new plan.
class RefineResult {
  const RefineResult({
    this.plan,
    this.reply,
    this.awaitingPaste = false,
    this.pendingSourceId,
    this.importedPlaceCount = 0,
    this.updatedDay,
  });

  final PlanRevision? plan;
  final String? reply;

  /// The source was blocked; the next message should be the pasted text.
  final bool awaitingPaste;
  final String? pendingSourceId;
  final int importedPlaceCount;
  final int? updatedDay;

  factory RefineResult.fromJson(Map<String, dynamic> json) {
    final planJson = json['plan'] as Map<String, dynamic>?;
    final importJson = json['import'] as Map<String, dynamic>?;
    return RefineResult(
      plan: planJson == null ? null : PlanRevision.fromJson(planJson),
      reply: json['reply'] as String?,
      awaitingPaste: json['awaiting_paste'] as bool? ?? false,
      pendingSourceId: json['pending_source_id'] as String?,
      importedPlaceCount: (importJson?['places'] as List?)?.length ?? 0,
      updatedDay: (json['updated_day'] as num?)?.toInt(),
    );
  }
}

File: server/src/schemas/plan.ts
import { z } from 'zod';

/**
 * The Part B contract from the Trip Planner spec.
 *
 * One deliberate addition: `conversational_summary` carries Part A (the prose
 * the user reads). The spec has the model emit prose *outside* a fenced JSON
 * block, which means parsing it back out of free text. Folding it into the
 * schema lets us constrain the whole response with structured outputs, so a
 * malformed response becomes near-impossible rather than a retry path. The
 * two-part contract is preserved — Part A is `conversational_summary`, Part B
 * is everything else.
 */

export const PaceSchema = z.enum(['packed', 'moderate', 'relaxed']);
export const TimeOfDaySchema = z.enum(['morning', 'afternoon', 'evening']);
export const PlanStatusSchema = z.enum(['complete', 'needs_info', 'infeasible']);

/**
 * Per-category budget. The Budget tab renders two stacked bars per category —
 * what was planned against what the itinerary actually estimates — so each
 * category carries both numbers rather than a single figure.
 */
export const CategoryBudgetSchema = z.object({
  planned: z.number(),
  estimated: z.number(),
});

export const BudgetBreakdownSchema = z.object({
  lodging: CategoryBudgetSchema,
  transport: CategoryBudgetSchema,
  food: CategoryBudgetSchema,
  activities: CategoryBudgetSchema,
  buffer: CategoryBudgetSchema,
});

export const TripSummarySchema = z.object({
  title: z.string(),
  destinations: z.array(z.string()),
  start_date: z.string().nullable(),
  end_date: z.string().nullable(),
  duration_days: z.number(),
  currency: z.string(),
  total_budget: z.number().nullable(),
  budget_breakdown: BudgetBreakdownSchema,
  estimated_total_cost: z.number(),
  over_budget: z.boolean(),
  assumptions: z.array(z.string()),
});

export const PlanMemberSchema = z.object({
  id: z.string(),
  name: z.string(),
  interests: z.array(z.string()),
  pace: PaceSchema,
  dietary_restrictions: z.array(z.string()),
  accessibility_needs: z.array(z.string()),
  deal_breakers: z.array(z.string()),
});

/** Categories the Group tab renders as a pill on each conflict card. */
export const ConflictTagSchema = z.enum([
  'Pace',
  'Mobility',
  'Food',
  'Budget',
  'Interests',
]);

export const ConflictSchema = z.object({
  tag: ConflictTagSchema,
  description: z.string(),
  members_involved: z.array(z.string()),
  resolution: z.string(),
});

export const BlockSchema = z.object({
  time_of_day: TimeOfDaySchema,
  activity: z.string(),
  description: z.string(),
  location: z.string(),
  estimated_duration_minutes: z.number(),
  estimated_cost_per_person: z.number().nullable(),
  suited_for_members: z.array(z.string()),
  optional: z.boolean(),
  weather_backup: z.string().nullable(),
});

export const DaySchema = z.object({
  day: z.number(),
  date: z.string().nullable(),
  location: z.string(),
  lodging_area_suggestion: z.string().nullable(),
  blocks: z.array(BlockSchema),
  notes: z.string().nullable(),
});

/**
 * A question the planner needs answered, shaped for the design's question
 * cards: the ask, the decision it unblocks, and how to answer it.
 */
export const ClarifyingQuestionSchema = z.object({
  id: z.string(),
  question: z.string(),
  /** What changes depending on the answer — rendered under the question. */
  why: z.string(),
  answer_type: z.enum(['choice', 'text']),
  /** Populated for 'choice'; empty for 'text'. */
  options: z.array(z.string()),
  /** Hint text for 'text' questions; null for 'choice'. */
  placeholder: z.string().nullable(),
});

export const PlanSchema = z.object({
  conversational_summary: z.string(),
  status: PlanStatusSchema,
  missing_info: z.array(z.string()),
  trip: TripSummarySchema,
  members: z.array(PlanMemberSchema),
  conflicts: z.array(ConflictSchema),
  itinerary: z.array(DaySchema),
  packing_and_prep_notes: z.array(z.string()),
  verify_before_booking: z.array(z.string()),
  clarifying_questions: z.array(ClarifyingQuestionSchema),
});

export type Plan = z.infer<typeof PlanSchema>;
export type PlanDay = z.infer<typeof DaySchema>;
export type PlanBlock = z.infer<typeof BlockSchema>;

/**
 * JSON Schema handed to the Messages API as `output_config.format`.
 *
 * Structured outputs require `additionalProperties: false` and an explicit
 * `required` list on every object. Nullable fields are expressed as a type
 * union rather than an optional key, so the model must always emit the key.
 */
const str = { type: 'string' } as const;
const num = { type: 'number' } as const;
const bool = { type: 'boolean' } as const;
const strArray = { type: 'array', items: str } as const;
const nullableStr = { type: ['string', 'null'] } as const;
const nullableNum = { type: ['number', 'null'] } as const;

const obj = <P extends Record<string, unknown>>(properties: P) => ({
  type: 'object' as const,
  properties,
  required: Object.keys(properties),
  additionalProperties: false as const,
});

const categoryBudget = (label: string) =>
  obj({
    planned: { ...num, description: `What the group intended to spend on ${label}.` },
    estimated: {
      ...num,
      description: `What this itinerary actually costs for ${label}.`,
    },
  });

export const PLAN_JSON_SCHEMA = obj({
  conversational_summary: {
    ...str,
    description:
      'PART A. Written in the first person, plainly, as someone reporting what they did: what you planned and why, assumptions you made, conflicts you resolved and how, and any budget problem. Two or three short paragraphs. Lead with what you decided, not a greeting. No promotional adjectives, no exclamation marks. Do not restate the itinerary day by day — the structured fields carry it.',
  },
  status: { type: 'string', enum: ['complete', 'needs_info', 'infeasible'] },
  missing_info: {
    ...strArray,
    description: 'Only non-empty when status is needs_info.',
  },
  trip: obj({
    title: str,
    destinations: strArray,
    start_date: { ...nullableStr, description: 'YYYY-MM-DD, or null if flexible' },
    end_date: nullableStr,
    duration_days: num,
    currency: str,
    total_budget: nullableNum,
    budget_breakdown: obj({
      lodging: categoryBudget('lodging'),
      transport: categoryBudget('transport'),
      food: categoryBudget('food'),
      activities: categoryBudget('activities'),
      buffer: categoryBudget('contingency'),
    }),
    estimated_total_cost: num,
    over_budget: bool,
    assumptions: {
      ...strArray,
      description:
        'Every fact you decided for yourself rather than being told — dates, budget, trip length, region, anything. One short entry each, naming what you chose: "Four days, since no length was given". Saying it in the summary does not cover this: the summary is prose the app cannot act on, and this array is what lets the group see and correct each guess. If you assumed nothing, leave it empty.',
    },
  }),
  members: {
    type: 'array',
    items: obj({
      id: str,
      name: str,
      interests: strArray,
      pace: { type: 'string', enum: ['packed', 'moderate', 'relaxed'] },
      dietary_restrictions: strArray,
      accessibility_needs: strArray,
      deal_breakers: strArray,
    }),
  },
  conflicts: {
    type: 'array',
    items: obj({
      tag: {
        type: 'string',
        enum: ['Pace', 'Mobility', 'Food', 'Budget', 'Interests'],
        description: 'The single category this conflict is mostly about.',
      },
      description: str,
      members_involved: strArray,
      resolution: str,
    }),
  },
  itinerary: {
    type: 'array',
    items: obj({
      day: num,
      date: nullableStr,
      location: str,
      lodging_area_suggestion: nullableStr,
      blocks: {
        type: 'array',
        items: obj({
          time_of_day: { type: 'string', enum: ['morning', 'afternoon', 'evening'] },
          activity: str,
          description: str,
          location: str,
          estimated_duration_minutes: num,
          estimated_cost_per_person: nullableNum,
          suited_for_members: {
            ...strArray,
            description: 'Member ids from the members array.',
          },
          optional: bool,
          weather_backup: nullableStr,
        }),
      },
      notes: nullableStr,
    }),
  },
  packing_and_prep_notes: strArray,
  verify_before_booking: {
    ...strArray,
    description: 'Time-sensitive things the user must verify before booking.',
  },
  clarifying_questions: {
    type: 'array',
    items: obj({
      id: { ...str, description: 'Short stable slug, e.g. "separate_arrivals".' },
      question: str,
      why: {
        ...str,
        description:
          'One sentence naming the planning decision this answer unblocks.',
      },
      answer_type: { type: 'string', enum: ['choice', 'text'] },
      options: {
        ...strArray,
        description: 'Choices when answer_type is "choice"; empty otherwise.',
      },
      placeholder: {
        ...nullableStr,
        description: 'Hint for "text" questions; null for "choice".',
      },
    }),
  },
});
